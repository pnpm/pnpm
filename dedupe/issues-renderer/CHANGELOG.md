# @pnpm/dedupe.issues-renderer

## 2.0.0

### Major Changes

- 43cdd87: Node.js v16 support dropped. Use at least Node.js v18.12.

### Patch Changes

- Updated dependencies [43cdd87]
  - @pnpm/dedupe.types@2.0.0

## 1.0.0

### Major Changes

- 6850bb135: Initial release.

### Patch Changes

- Updated dependencies [6850bb135]
  - @pnpm/dedupe.types@1.0.0
