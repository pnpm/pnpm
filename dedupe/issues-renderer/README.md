# @pnpm/dedupe.issues-renderer

> Visualizes "pnpm dedupe --check" issues

## Installation

```
pnpm add @pnpm/dedupe.issues-renderer
```

## License

[MIT](LICENSE)
