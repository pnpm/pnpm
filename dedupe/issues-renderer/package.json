{
  "name": "@pnpm/dedupe.issues-renderer",
  "version": "2.0.0",
  "description": "Visualize pnpm dedupe --check issues.",
  "bugs": {
    "url": "https://github.com/pnpm/pnpm/issues"
  },
  "main": "lib/index.js",
  "types": "lib/index.d.ts",
  "files": [
    "lib",
    "!*.map"
  ],
  "keywords": [
    "pnpm9"
  ],
  "license": "MIT",
  "engines": {
    "node": ">=18.12"
  },
  "repository": "https://github.com/pnpm/pnpm/blob/main/dedupe/issues-renderer",
  "scripts": {
    "_test": "jest",
    "test": "pnpm run compile && pnpm run _test",
    "lint": "eslint \"src/**/*.ts\" \"test/**/*.ts\"",
    "prepublishOnly": "pnpm run compile",
    "compile": "tsc --build && pnpm run lint --fix"
  },
  "homepage": "https://github.com/pnpm/pnpm/blob/main/dedupe/issues-renderer#readme",
  "funding": "https://opencollective.com/pnpm",
  "dependencies": {
    "@pnpm/dedupe.types": "workspace:*",
    "archy": "^1.0.0",
    "chalk": "^4.1.2"
  },
  "devDependencies": {
    "@pnpm/dedupe.issues-renderer": "workspace:*",
    "@types/archy": "0.0.33",
    "strip-ansi": "^6.0.1"
  },
  "exports": {
    ".": "./lib/index.js"
  }
}
