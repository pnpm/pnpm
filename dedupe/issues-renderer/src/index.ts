import {
  type ResolutionChange,
  type DedupeCheckIssues,
  type ResolutionChangesByAlias,
  type SnapshotsChanges,
} from '@pnpm/dedupe.types'
import archy from 'archy'
import chalk from 'chalk'

export function renderDedupeCheckIssues (dedupeCheckIssues: DedupeCheckIssues): string {
  const importersReport = report(dedupeCheckIssues.importerIssuesByImporterId)
  const packagesReport = report(dedupeCheckIssues.packageIssuesByDepPath)

  const lines = []
  if (importersReport !== '') {
    lines.push(chalk.blueBright.underline('Importers'))
    lines.push(importersReport)
    lines.push('')
  }
  if (packagesReport !== '') {
    lines.push(chalk.blueBright.underline('Packages'))
    lines.push(packagesReport)
    lines.push('')
  }

  return lines.join('\n')
}

/**
 * Render snapshot changes. Expected to return an empty string for no changes.
 */
function report (snapshotChanges: SnapshotsChanges): string {
  return [
    ...Object.entries(snapshotChanges.updated).map(([alias, updates]) => archy(toArchy(alias, updates))),
    ...snapshotChanges.added.map((id) => `${chalk.green('+')} ${id}`),
    ...snapshotChanges.removed.map((id) => `${chalk.red('-')} ${id}`),
  ].join('\n')
}

function toArchy (name: string, issue: ResolutionChangesByAlias): archy.Data {
  return {
    label: name,
    nodes: Object.entries(issue).map(([alias, change]) => toArchyResolution(alias, change)),
  }
}

function toArchyResolution (alias: string, change: ResolutionChange): archy.Data {
  switch (change.type) {
  case 'added':
    return { label: `${chalk.green('+')} ${alias} ${chalk.gray(change.next)}` }
  case 'removed':
    return { label: `${chalk.red('-')} ${alias} ${chalk.gray(change.prev)}` }
  case 'updated':
    return { label: `${alias} ${chalk.red(change.prev)} ${chalk.gray('â†’')} ${chalk.green(change.next)}` }
  }
}
