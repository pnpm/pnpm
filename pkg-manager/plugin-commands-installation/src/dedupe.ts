import { docsUrl } from '@pnpm/cli-utils'
import { OPTIONS, UNIVERSAL_OPTIONS } from '@pnpm/common-cli-options-help'
import { dedupeDiffCheck } from '@pnpm/dedupe.check'
import renderHelp from 'render-help'
import { type InstallCommandOptions, rcOptionsTypes as installCommandRcOptionsTypes } from './install'
import { installDeps } from './installDeps'
import omit from 'ramda/src/omit'

// In general, the "pnpm dedupe" command should use .npmrc options that "pnpm install" would also accept.
export function rcOptionsTypes (): Record<string, unknown> {
  // Some options on pnpm install (like --frozen-lockfile) don't make sense on pnpm dedupe.
  return omit(['frozen-lockfile'], installCommandRcOptionsTypes())
}

export function cliOptionsTypes (): Record<string, unknown> {
  return {
    ...rcOptionsTypes(),
    check: Boolean,
  }
}

export const commandNames = ['dedupe']

export function help (): string {
  return renderHelp({
    description: 'Perform an install removing older dependencies in the lockfile if a newer version can be used.',
    descriptionLists: [
      {
        title: 'Options',
        list: [
          ...UNIVERSAL_OPTIONS,
          {
            description: 'Check if running dedupe would result in changes without installing packages or editing the lockfile. Exits with a non-zero status code if changes are possible.',
            name: '--check',
          },
          OPTIONS.ignoreScripts,
          OPTIONS.offline,
          OPTIONS.preferOffline,
          OPTIONS.storeDir,
          OPTIONS.virtualStoreDir,
          OPTIONS.globalDir,
        ],
      },
    ],
    url: docsUrl('dedupe'),
    usages: ['pnpm dedupe'],
  })
}

export interface DedupeCommandOptions extends InstallCommandOptions {
  readonly check?: boolean
}

export async function handler (opts: DedupeCommandOptions): Promise<void> {
  const include = {
    dependencies: opts.production !== false,
    devDependencies: opts.dev !== false,
    optionalDependencies: opts.optional !== false,
  }
  return installDeps({
    ...opts,
    dedupe: true,
    include,
    includeDirect: include,
    lockfileCheck: opts.check ? dedupeDiffCheck : undefined,
  }, [])
}
