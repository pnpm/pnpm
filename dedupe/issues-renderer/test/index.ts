import { renderDedupeCheckIssues } from '@pnpm/dedupe.issues-renderer'
import stripAnsi from 'strip-ansi'

describe('renderDedupeCheckIssues', () => {
  test('prints removed packages and updated resolutions', () => {
    expect(stripAnsi(renderDedupeCheckIssues({
      importerIssuesByImporterId: {
        added: [],
        removed: [],
        updated: {
          '.': {
            '@types/node': { type: 'updated', prev: '14.18.42', next: '18.15.11' },
          },
        },
      },
      packageIssuesByDepPath: {
        added: [],
        removed: ['/@types/node/14.18.42'],
        updated: {
          '@types/tar-stream/2.2.2': {
            '@types/node': { type: 'updated', prev: '14.18.42', next: '18.15.11' },
          },
          '@types/tar/6.1.4': {
            '@types/node': { type: 'updated', prev: '14.18.42', next: '18.15.11' },
          },
        },
      },
    }))).toMatchSnapshot()
  })

  test('prints new importers and packages', () => {
    expect(stripAnsi(renderDedupeCheckIssues({
      importerIssuesByImporterId: {
        added: ['packages/a'],
        removed: [],
        updated: {
          '.': {
            'packages/a': { type: 'added', next: '0.0.0' },
          },
        },
      },
      packageIssuesByDepPath: {
        added: [
          // Suppose packages/a added a new @types/node dependency on 18.x.
          '/@types/node/18.15.11',
        ],
        removed: ['/@types/node/14.18.42'],
        updated: {
          '@types/tar-stream/2.2.2': {
            '@types/node': { type: 'updated', prev: '14.18.42', next: '18.15.11' },
          },
          '@types/tar/6.1.4': {
            '@types/node': { type: 'updated', prev: '14.18.42', next: '18.15.11' },
          },
        },
      },
    }))).toMatchSnapshot()
  })

  test('prints removed importer', () => {
    expect(stripAnsi(renderDedupeCheckIssues({
      importerIssuesByImporterId: {
        added: [],
        removed: ['packages/a'],
        updated: {
          '.': {
            '@types/node': { type: 'updated', prev: '18.15.11', next: '14.18.42' },
          },
        },
      },
      packageIssuesByDepPath: {
        added: ['/@types/node/14.18.42'],
        removed: ['/@types/node/18.15.11'],
        updated: {
          '@types/tar-stream/2.2.2': {
            '@types/node': { type: 'updated', prev: '18.15.11', next: '14.18.42' },
          },
          '@types/tar/6.1.4': {
            '@types/node': { type: 'updated', prev: '18.15.11', next: '14.18.42' },
          },
        },
      },
    }))).toMatchSnapshot()
  })
})
