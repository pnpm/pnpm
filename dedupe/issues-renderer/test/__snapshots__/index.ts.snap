// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderDedupeCheckIssues prints new importers and packages 1`] = `
"Importers
.
└── + packages/a 0.0.0

+ packages/a

Packages
@types/tar-stream/2.2.2
└── @types/node 14.18.42 → 18.15.11

@types/tar/6.1.4
└── @types/node 14.18.42 → 18.15.11

+ /@types/node/18.15.11
- /@types/node/14.18.42
"
`;

exports[`renderDedupeCheckIssues prints removed importer 1`] = `
"Importers
.
└── @types/node 18.15.11 → 14.18.42

- packages/a

Packages
@types/tar-stream/2.2.2
└── @types/node 18.15.11 → 14.18.42

@types/tar/6.1.4
└── @types/node 18.15.11 → 14.18.42

+ /@types/node/14.18.42
- /@types/node/18.15.11
"
`;

exports[`renderDedupeCheckIssues prints removed packages and updated resolutions 1`] = `
"Importers
.
└── @types/node 14.18.42 → 18.15.11


Packages
@types/tar-stream/2.2.2
└── @types/node 14.18.42 → 18.15.11

@types/tar/6.1.4
└── @types/node 14.18.42 → 18.15.11

- /@types/node/14.18.42
"
`;
